alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,108 @@
+        Self { damaged_lines, size_info }
+    }
+
+    #[inline]
+    fn rect_for_line(&self, line_damage: LineDamageBounds) -> Rect {
+        let size_info = &self.size_info;
+        let y_top = size_info.height() - size_info.padding_y();
+        let x = size_info.padding_x() + line_damage.left as u32 * size_info.cell_width();
+        let y = y_top - (line_damage.line + 1) as u32 * size_info.cell_height();
+        let width = (line_damage.right - line_damage.left + 1) as u32 * size_info.cell_width();
+        Rect { x, y, height: size_info.cell_height(), width }
+    }
+
+    // Make sure to damage near cells to include wide chars.
+    #[inline]
+    fn overdamage(&self, mut rect: Rect) -> Rect {
+        let size_info = &self.size_info;
+        rect.x = rect.x.saturating_sub(size_info.cell_width());
+        rect.width = cmp::min(size_info.width() - rect.x, rect.width + 2 * size_info.cell_width());
+        rect.y = rect.y.saturating_sub(size_info.cell_height() / 2);
+        rect.height = cmp::min(size_info.height() - rect.y, rect.height + size_info.cell_height());
+
+        rect
+    }
+}
+
+impl<'a> Iterator for RenderDamageIterator<'a> {
+    type Item = Rect;
+
+    fn next(&mut self) -> Option<Rect> {
+        let line = self.damaged_lines.next()?;
+
+        let mut total_damage = self.rect_for_line(line);
+
+        // Currently merged rect and its area.
+        let mut merged_rect = total_damage;
+        let mut merged_area = rect_area(&merged_rect);
+
+        // Iterator to look rects ahead and try to form larger rect without advancing the orignal
+        // iterator.
+        let mut damaged_lines = self.damaged_lines.clone();
+
+        while let Some(line) = damaged_lines.next() {
+            let next_rect = self.rect_for_line(line);
+            let next_rect_area = rect_area(&next_rect);
+            let next_merged_rect = merge_rects(merged_rect, next_rect);
+            let next_merged_area = rect_area(&next_merged_rect) as f32;
+            let merged_ratio = (merged_area + next_rect_area) as f32 / next_merged_area;
+
+            // The approach is to minimize amount of damaged rects we report to compositor, not to
+            // achieve the best accuracy, since the best accuracy is achieved if you don't merge
+            // anything. So we try to merge rects if the area ratio is higher than
+            // `NIM_DAMAGED_AREA_RATIO`, if it's lower that that ratio, but higher than
+            // `MIN_DAMAGED_AREA_RATIO_ERR` we try to merge more rectangles until we reach
+            // `MIN_DAMAGED_AREA_RATIO`. If we haven't been able to reach `MIN_DAMAGED_AREA_RATIO`
+            // we abort merging and reporting what've collected so far.
+            if merged_ratio >= MIN_DAMAGED_AREA_RATIO {
+                total_damage = next_merged_rect;
+                // Update iterator position.
+                self.damaged_lines = damaged_lines.clone();
```

Comment:
Do you want to have `advance` approach back? Cloning iterator is pretty cheap, since you just clone offset into underlying collection and pointer to that collection. I thought that it's easier than having more extra variables for no reason.

Reply:
I don't really like either of these approaches.